**Create Matrix A, X, Eta,B**

In [1]:
import random
import numpy as np
import math
import time

A = [[0] * 500] * 1000
for row in range(1000):
    for col in range(500):
        A[row][col] = np.random.uniform(-1, 1)

X = [0] * 500
for i in range(500):
    X[i] = np.random.uniform(-1, 1)
 
sd = math.pow(0.5, 0.5)
E= np.random.normal(0, sd, 1000)


B = np.add(np.matmul(A, X),E)
    

**Loss Calculation**

In [2]:
def loss(a, w, b):
    final_loss = 0
    temp= np.matmul(a, w)-b
    final_loss = np.matmul(np.transpose(temp), temp)
    return final_loss

**Implementing Gardiant Descent**

In [3]:
def gradient_descent(A, B,learning_rate, iterations):
    w = [0]*500
    print("Starting Loss:", loss(A, w, B))
    for i in range(iterations):
        w = w -learning_rate * gradient(A, w, B)
        print("loss at "+str(i)+"th iteration " +str(loss(A, w, B)))
    return w


def gradient(A, w, B):
    grad = 2 * (np.matmul(np.matmul(np.transpose(A), A),w)-np.matmul(np.transpose(A), B))
    return grad



**Calling Gradient Descent , Finding distance , Time**

In [4]:
GD_start = time.time()
Final_Weights= gradient_descent(A, B, 0.000001, 50)
GD_end = time.time()

Starting Loss: 9486.390772817993
loss at 0th iteration 4845.928518540698
loss at 1th iteration 2597.4319894455903
loss at 2th iteration 1507.9421144460287
loss at 3th iteration 980.0390590633931
loss at 4th iteration 724.2481221513503
loss at 5th iteration 600.3067985707992
loss at 6th iteration 540.2520840592332
loss at 7th iteration 511.15308276672414
loss at 8th iteration 497.0534091089903
loss at 9th iteration 490.2215321226911
loss at 10th iteration 486.9112042180218
loss at 11th iteration 485.3072129783656
loss at 12th iteration 484.5300126613624
loss at 13th iteration 484.1534268560189
loss at 14th iteration 483.97095540601606
loss at 15th iteration 483.88254041640766
loss at 16th iteration 483.83969968495427
loss at 17th iteration 483.8189415727882
loss at 18th iteration 483.8088834060371
loss at 19th iteration 483.8040098068589
loss at 20th iteration 483.8016483458155
loss at 21th iteration 483.8005041199457
loss at 22th iteration 483.7999496950294
loss at 23th iteration 483.7

In [5]:
print("Final Weights Matrix:")
print(Final_Weights)

Final Weights Matrix:
[-1.22185995e-02  1.82789021e-02 -1.27110873e-03  5.65557575e-03
  1.44905522e-02  9.14305893e-03  1.10329087e-02  1.52718859e-02
 -1.13991157e-02  1.03909208e-03  8.86237605e-03 -8.05666573e-03
 -2.81467278e-03  4.19630212e-04  1.31536810e-02  7.04138350e-03
 -1.66401678e-02 -7.65585029e-03  5.50185943e-03 -4.60152244e-03
 -1.15397916e-03  7.25366178e-03 -1.16848306e-02 -9.27344370e-03
 -1.74674694e-02 -7.76091762e-03 -1.03391392e-02  1.86988372e-02
 -2.41677197e-03  4.96945496e-03  1.10227287e-02  1.28463103e-02
 -1.91835636e-02 -1.12576991e-02 -4.50800299e-03  1.63700526e-02
  1.51820973e-02 -1.36839044e-02  1.20201553e-02 -9.36001585e-03
 -6.98954236e-03  3.63936265e-03  1.75125972e-02 -1.62643796e-02
  5.99493555e-03 -1.64834140e-02  1.44924087e-02  1.61472698e-02
  8.79617239e-04 -2.22718378e-03  4.61391536e-03 -2.70624465e-03
  6.29546153e-04  1.13152133e-02  1.18149473e-02 -1.85573342e-02
 -1.20515965e-02 -1.37868179e-03  1.91941919e-02 -1.15643269e-02
 -7

In [6]:
print("Time Taken for gradient descent",GD_end-GD_start)
temp=X-Final_Weights
dist=np.matmul(np.transpose(temp),temp)
print("Distance to the hidden X taken:",dist)

Time Taken for gradient descent 5.80753493309021
Distance to the hidden X taken: 159.55118166228192


**Comparision with Weights matrix that is found using the Regression Formula**

In [7]:
Reg_start = time.time()
Final_reg_weights=np.matmul(np.matmul(np.linalg.inv(np.matmul(np.transpose(A),A)),np.transpose(A)),B)
Reg_end = time.time()

In [8]:
print("Time Taken for Regression Formula:",Reg_end-Reg_start)

Time Taken for Regression Formula: 0.10427212715148926


# Observations

<br>1. The Function Value after 50 steps =484
<br>2. The distance to the Hidden X =160
<br>3. The time taken with gradiant descent is 58 times the time taken using formuls.The ratio will increse more if number of iterations is increased in gradient descent.
<br>
<br> Here we had used the learning rate =0.000001 , beacuse if we take learning rate =0.1 and start running the gradient descent with intitial weight that is all zeros.Since the intial weight already very close to original X and if we calculate the gradient with learning rate =0.1 the algorithm pushing the new X that is really far and Function value is increasing . So with 0.1 learning rate the algorithm is not learning , instead producing high errors with oscillating between minimum.
<br>By trying various learning rated i found that leaning_rate=0.000001 is optimal and reduced Function value really well.